chain/chain/src/resharding/trie_state_resharder.rs

```diff
@@ -0,0 +1,260 @@
+
+use borsh::{BorshDeserialize, BorshSerialize};
+use near_primitives::hash::CryptoHash;
+use near_store::adapter::trie_store::TrieStoreUpdateAdapter;
+use near_store::adapter::{StoreAdapter, StoreUpdateAdapter};
+use near_store::db::TRIE_STATE_RESHARDING_STATUS_KEY;
+use near_store::metrics::trie_state_metrics;
+use near_store::{DBCol, ShardTries, StorageError};
+
+use crate::resharding::event_type::ReshardingSplitShardParams;
+use crate::types::RuntimeAdapter;
+use near_chain_configs::{MutableConfigValue, ReshardingConfig, ReshardingHandle};
+use near_chain_primitives::Error;
+use near_o11y::metrics::IntGauge;
+use near_primitives::shard_layout::ShardUId;
+
+#[derive(BorshSerialize, BorshDeserialize)]
+struct TrieStateReshardingChildStatus {
+    shard_uid: ShardUId,
+    state_root: CryptoHash,
+    next_key: Vec<u8>,
+
+    #[borsh(skip)]
+    metrics: Option<TrieStateResharderMetrics>,
+}
+
+impl TrieStateReshardingChildStatus {
+    fn new(shard_uid: ShardUId, state_root: CryptoHash) -> Self {
+        Self { shard_uid, state_root, next_key: vec![], metrics: None }
+    }
+}
+
+#[derive(BorshSerialize, BorshDeserialize)]
+struct TrieStateReshardingStatus {
+    shard_uid: ShardUId,
+    children: Vec<TrieStateReshardingChildStatus>,
+}
+
+impl TrieStateReshardingStatus {
+    fn new(
+        shard_uid: ShardUId,
+        left: TrieStateReshardingChildStatus,
+        right: TrieStateReshardingChildStatus,
+    ) -> Self {
+        Self { shard_uid, children: vec![left, right] }
+    }
+
+    fn with_metrics(mut self) -> Self {
+        for child in &mut self.children {
+            child.metrics = Some(TrieStateResharderMetrics::new(&child.shard_uid));
+        }
+        self
+    }
+
+    fn done(&self) -> bool {
+        self.children.is_empty()
+    }
+}
+
+/// TrieStateResharder is responsible for handling state resharding operations.
+pub struct TrieStateResharder {
+    runtime: Arc<dyn RuntimeAdapter>,
+    /// Controls cancellation of background processing.
+    pub handle: ReshardingHandle,
+    /// Configuration for resharding.
+    resharding_config: MutableConfigValue<ReshardingConfig>,
+}
+
+impl TrieStateResharder {
+    pub fn new(
+        runtime: Arc<dyn RuntimeAdapter>,
+        handle: ReshardingHandle,
+        resharding_config: MutableConfigValue<ReshardingConfig>,
+    ) -> Self {
+        Self { runtime, handle, resharding_config }
+    }
+
+    // Processes one batch of a trie state resharding and updates the status,
+    // also persisting the status to the store.
+    fn process_batch_and_update_status(
+        &self,
+        status: &mut TrieStateReshardingStatus,
+    ) -> Result<(), Error> {
+        let batch_size = self.resharding_config.get().batch_size.as_u64() as usize;
+        let batch_delay = self.resharding_config.get().batch_delay.unsigned_abs();
+        while let Some(child) = status.children.first_mut() {
+            let mut store_update = self.runtime.store().store_update();
```

Comment:
It could be useful to wrap the code between this line and the `commit()` inside a tracing block (a span, basically) , to know how long it takes

Reply:
also moving the sleep before declaring the span